### PR TITLE
*: allow different shard_row_id prefixes in one transaction

### PR DESCRIPTION
**Pull request by bobotu**
### What problem does this PR solve?
Shard will use a same prefix for each rowid in one transaction, this will cause problem when committing a large transaction.

### What is changed and how it works?
Add a `tidb_shard_allocate_step` variables to break this allocation into smaller steps. The default value is `math.MaxInt64`, so add it will not change the default behavior.

```sql
create table ttt(a int) SHARD_ROW_ID_BITS=10;
set @@tidb_shard_allocate_step=3;
insert into ttt values (1), (2), (3), (4), (5), (6), (7), (8), (9), (10), (11);

select _tidb_rowid, a from ttt order by a;
+---------------------+------+
| _tidb_rowid         | a    |
+---------------------+------+
| 3071454945866678273 |    1 |
| 3071454945866678274 |    2 |
| 3071454945866678275 |    3 |
| 6611284252979888132 |    4 |
| 6611284252979888133 |    5 |
| 6611284252979888134 |    6 |
| 4017210867614482439 |    7 |
| 4017210867614482440 |    8 |
| 4017210867614482441 |    9 |
| 1927540640514572298 |   10 |
| 1927540640514572299 |   11 |
+---------------------+------+
```

### Check List 

Tests 

- Unit test

### Release note 

- Add `tidb_shard_allocate_step` to control max size of continuous rowid shard in one transaction.

### CODE REVIEW

File: table/tables/tables.go [hunk 1]
```diff
@@ -1153,12 +1152,10 @@ func allocHandleIDs(ctx sessionctx.Context, t table.Table, n uint64) (int64, int
 			return 0, 0, autoid.ErrAutoincReadFailed
 		}
 		txnCtx := ctx.GetSessionVars().TxnCtx
-		if txnCtx.Shard == nil {
-			shard := CalcShard(meta.ShardRowIDBits, txnCtx.StartTS, autoid.RowIDBitLength, true)
-			txnCtx.Shard = &shard
-		}
-		base |= *txnCtx.Shard
-		maxID |= *txnCtx.Shard
+		// shard := CalcShard(meta.ShardRowIDBits, txnCtx.StartTS, autoid.RowIDBitLength, true)
```
Comment (tangenta) [inline table/tables/tables.go:1155]:
```suggestion
```

File: table/tables/tables.go [hunk 2]
```diff
@@ -570,6 +568,7 @@ func (t *TableCommon) AddRecord(ctx sessionctx.Context, r []types.Datum, opts ..
 			// Make the IDs continuous benefit for the performance of TiKV.
 			stmtCtx := ctx.GetSessionVars().StmtCtx
 			stmtCtx.BaseRowID, stmtCtx.MaxRowID, err = allocHandleIDs(ctx, t, uint64(opt.ReserveAutoID))
+			println("reverse", stmtCtx.BaseRowID, stmtCtx.MaxRowID)
```
Comment (tangenta) [inline table/tables/tables.go:571]:
```suggestion
```

File: sessionctx/variable/session.go [hunk 3]
```diff
@@ -160,6 +168,29 @@ type TransactionContext struct {
 	LockExpire     uint32
 }
 
+func (tc *TransactionContext) GetShard(shardRowIDBits uint64, typeBitsLength uint64, reserveSignBit bool, count int) int64 {
+	if tc.shardRand == nil {
```
Comment (tangenta) [inline sessionctx/variable/session.go:172]:
```suggestion
	if shardRowIDBits == 0 {
		return 0
	}
	if tc.shardRand == nil {
```

**Comment by bobotu** [comment]
@tangenta PTAL

**Pull request merged**